src/model.jl

```diff
@@ -1,3 +1,90 @@
+"""
+    GenericModel{F}
+
+State wrapper to hold `Libtask.CTask` model initiated from `f`
+"""
+struct GenericModel{F} <: AbstractMCMC.AbstractModel
+    f::F
+    ctask::Libtask.TapedTask{F}
+end
+
+GenericModel(f, args...) = GenericModel(f, Libtask.TapedTask(f, args...))
+Base.copy(model::GenericModel) = GenericModel(model.f, copy(model.ctask))
+
+"""
+    Trace{F,R}
+"""
+struct Trace{F,R}
+    model::F
+    rng::R
+
+    Trace(model::F, rng::R) where {F,R} = new{F,R}(model, rng) # Bring back the default since we override
+
+    function Trace(model::F, rng::R) where {F<:GenericModel,R}
+        # add backward reference
+        newtrace = new{F,R}(model, rng)
+        addreference!(model.ctask.task, newtrace)
+        return newtrace
+    end
+end
+
+const Particle = Trace
+
+const SSMTrace{R} = Trace{<:AbstractStateSpaceModel,R}
+const GenericTrace{R} = Trace{<:GenericModel,R}
+
+# Copy task
+Base.copy(trace::GenericTrace) = Trace(copy(trace.model), deepcopy(trace.rng))
+
+# create a backward reference in task_local_storage
+function addreference!(task::Task, trace::GenericTrace)
+    if task.storage === nothing
+        task.storage = IdDict()
+    end
+    task.storage[:__trace] = trace
+
+    return task
+end
+
+current_trace() = current_task().storage[:__trace]
+
+# Task copying version of fork for Trace.
+function fork(trace::GenericTrace, isref::Bool=false)
+    newtrace = copy(trace)
+    isref && delete_retained!(newtrace.model.f)
+
+    # add backward reference
+    addreference!(newtrace.model.ctask.task, newtrace)
+    return newtrace
+end
+
+# PG requires keeping all randomness for the reference particle
+# Create new task and copy randomness
+function forkr(trace::GenericTrace)
+    newf = reset_model(trace.model.f)
+    Random123.set_counter!(trace.rng, 1)
+
+    ctask = Libtask.TapedTask(newf, trace.rng)
+    new_tapedmodel = GenericModel(newf, ctask)
+
+    # add backward reference
+    newtrace = Trace(new_tapedmodel, trace.rng)
+    addreference!(ctask.task, newtrace)
+    gen_refseed!(newtrace)
+    return newtrace
+end
+
+function gen_refseed!(part::Particle)
+    seed = split(state(part.rng.rng), 1)
+    return set_refseed!(part.rng, seed[1])
```

Comment:
Would this line override the last `key` in `part.rng::TracedRNG`? If so, will it cause issues during replaying particle `part`? 
```suggestion
    return set_refseed!(part.rng, seed[1])
```

Reply:
I'm hacking around a bit here, but in the current implementation we store T+1 seeds (for a model with T steps) because we iterate T+1 times the model (to get the final un-weighted sample). This last seed is not used in subsequent runs. 

We could also just append a seed to the reference particle here or track it as an optional field in `TracedRNG` here:
https://github.com/TuringLang/AdvancedPS.jl/blob/02a5258c12b2db7f0dc6635311349465ce6d23eb/src/model.jl#L63-L75

Reply:
I like tracking it as an optional field in `TracedRNG`. We can simply leave it as uninitialised for non-reference particles. Such a design can improve the clarity, which we should aim for.

Reply:
Agree, it was really hard to follow the logic. Even if I wrote it ...